### PR TITLE
Add exponential decay for learning rate.

### PR DESCRIPTION
**Pull request by gtoderici**
This enables users to specify an exponential decay of the learning rate, and adds a LSTM model. By default it decays by .95 every 4M steps now.

### CODE REVIEW

**Comment by gtoderici** [comment]
I signed.

**Comment by LeegleechN** [comment]
Have you checked the effect this has on the Logistic, Moe, and DBoF models?

**Comment by gtoderici** [comment]
It should affect every model in some way. I could default it to once per
8M, if you wish.

**Comment by gtoderici** [comment]
I want to clarify further what this does: basically it adds some basic functionality which is missing. All of it can be "disabled" by setting the appropriate values. This functionality should be beneficial to all models (both learning rate decay and clipping). 

Note: I haven't had the chance to test all. I only tested the video-level Logistic model, but if we want to be thorough, we may want to have different schedules for each model. I think that figuring out the correct hyperparameters for each model is beyond the scope of this request. 

Please let me know what you think.

**Comment by LeegleechN** [comment]
I checked the gradient magnitudes for my Moe-like model, and they are two orders of magnitude under the clipping threshold of 1.0. So the only effect that would have on similar models would be to slow down the training slightly.

The gradient magnitudes also go down over time, presumably due to the use of Adam. So we are getting some effect of a decaying learning rate already, although it doesn't have a stairstep pattern.

**Comment by gtoderici** [comment]
I ran a baseline vs. this change for MoE and didn't see any change learning, except that the version with clipped gradients seemed to learn faster. If you'd like to be "safe" (or the *opposite* of safe, given that this is a fraction of the global norm?), I could default this to a large value.

Similarly, I can make the decay steps a larger value. Again, even with Adam, this has an effect.
